rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -835,14 +844,13 @@ def train(
                 optimizer=tf.keras.optimizers.Adam(self.component_config[LEARNING_RATE])
             )
 
-        data_generator, validation_data_generator = train_utils.create_data_generators(
-            model_data,
-            self.component_config[BATCH_SIZES],
-            self.component_config[EPOCHS],
-            self.component_config[BATCH_STRATEGY],
-            self.component_config[EVAL_NUM_EXAMPLES],
-            self.component_config[RANDOM_SEED],
+        data_generator, validation_data_generator = (
+            MessageStreamDataGenerator(
```

Comment:
do you know if the `Caution:` part from [here](https://www.tensorflow.org/guide/data#consuming_python_generators)  applies?

Reply:
Yeah maybe something like Dataset.from_generator would be the right way to go

Reply:
so it seems `tf.data` is preferred for non-python preprocessing (== all pre-processing is expressed via TensorFlow functions) while the Keras sequence is preferred if you do a lot of Python preprocessing. So in our case the Keras sequence seems to be preferable.

Reply:
> Yeah maybe something like Dataset.from_generator would be the right way to go

Constructing a `tf.data.Dataset` from `Dataset.generator` becomes cumbersome when you need control on what should happen after the end of every epoch. If you need that control you need to write a custom `train_step` function instead of replying on `keras.model.fit()` to handle that for you. 

[tf.keras.utils.Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence#on_epoch_end) explicitly has a function for it. That's why we shifted from `tf.data.Dataset.from_generator` to `tf.keras.utils.Sequence` during the transition from [pure TF to reusing Keras methods](https://github.com/RasaHQ/rasa/pull/7438).

However, we recently recognized that tensorflow has [breaking support for `tf.RaggedTensors` inside `tf.keras.utils.Sequence`.](https://github.com/tensorflow/tensorflow/issues/48188#issue-844491927). 
Using `tf.RaggedTensors` will simplify a LOT of code inside our TF models. So, I am all up for exploring how we can use `tf.data.Dataset.from_generator` without having to write our own `model.fit()` function.